vdr/doc-updater.go

```diff
@@ -0,0 +1,111 @@
+package vdr
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"encoding/base64"
+	"errors"
+
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/nuts-foundation/go-did"
+
+	nutsCrypto "github.com/nuts-foundation/nuts-node/crypto"
+)
+
+// NutsDocUpdater contains helper methods to update a nuts document.
+type NutsDocUpdater struct {
+	// keyCreator is used for getting a fresh key and use it to generate the Nuts DID
+	keyCreator nutsCrypto.KeyCreator
+}
+
+// newNamingFnForExistingDID returns a KIDNamingFunc that can be used as param in the keyCreator.New function.
+// It wraps the KIDNamingFunc with the context of the DID of the document.
+// It returns a keyID in the form of the documents DID with the new keys thumbprint as fragment.
+func newNamingFnForExistingDID(existingDID did.DID) nutsCrypto.KIDNamingFunc {
+	return func(pKey crypto.PublicKey) (string, error) {
+		ecPKey, ok := pKey.(*ecdsa.PublicKey)
+		if !ok {
+			return "", errors.New("could not generate kid: invalid key type")
+		}
+
+		if ecPKey.Curve == nil {
+			return "", errors.New("could not generate kid: empty key curve")
+		}
+
+		jwKey, err := jwk.New(pKey)
+		if err != nil {
+			return "", err
+		}
+		thumbprint, err := jwKey.Thumbprint(thumbprintAlg)
+		if err != nil {
+			return "", err
+		}
+
+		existingDID.Fragment = base64.RawURLEncoding.EncodeToString(thumbprint)
+
+		return existingDID.String(), nil
+	}
+}
+
+// AddNewAuthenticationMethodToDIDDocument creates a new VerificationMethod of type JsonWebKey2020 with a freshly generated key
+func (u NutsDocUpdater) AddNewAuthenticationMethodToDIDDocument(doc *did.Document) error {
```

Comment:
Where is this used?

Reply:
We should move this to another abstraction layer. Added references to #123